### ISSUE TITLE
Why sftp.scandir checks for str?

### ISSUE THREAD

**Issue by Mulugruntz** [ronf/asyncssh#350]
Hi!

I saw this here https://github.com/ronf/asyncssh/blob/875330da4bb0322d872f702dbb1f44c7e6137c48/asyncssh/sftp.py#L3223

And I'm wondering what was the rationale behind this choice?

```python
from pathlib import Path
import asyncio
import asyncssh

my_dir = Path("/", "home", "debian", "path", "there")
url, user, password = 'ssh.foo.bar', 'user', 'pass'

async def main():
        async with asyncssh.connect(url, username=user, password=password) as conn:
            async with conn.start_sftp_client() as sftp:
                async for f in sftp.scandir(my_dir):
                    fn = my_dir / f.filename # fails
                async for f in sftp.scandir(str(my_dir)):
                    fn = my_dir / f.filename # works ok

if __name__ == "__main__":
    asyncio.run(main())
```

With the example above, it doesn't look very API-ish to have different behaviours depending on the input type (especially as it may fail way later in the code).
I would guess there's a reason but I can't think of any.

**Comment by ronf** [comment]
This code was added before support for passing in pathlib paths was added. Its original intention was to allow you to pass in paths which were either of type `str` or of type `bytes`. On the wire, `str` paths are always converted to `bytes` (using UTF-8 encoding), but to make things easier for the caller, this could would convert the returned `bytes` values back to `str` automatically when a `str` path was passed in.

Now that paths are supported, this code should probably be a 3-way check, returning `bytes`, `str`, or `PurePath` values depending on what was passed in. Alternately, as is done in some other places, I could have it return `str` values when either `str` or `Path` values are passed in, which should be enough to solve the issue you raised here.

It's important to support `bytes` values in case someone has a filesystem which encodes Unicode characters in filenames using something other than UTF-8, and in that case it's also important to return the data without trying to decode it back to a string, since the encoding used is not known.

**Comment by ronf** [comment]
As you noted here, you can work around the problem by converting the `Path` value to `str` for now, but I'll work on fixing it so you don't have to. Thanks for the report!

**Comment by ronf** [comment]
The following patch should fix this:

```patch
diff --git a/asyncssh/sftp.py b/asyncssh/sftp.py
index 117f69b..b66c373 100644
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -3220,7 +3220,7 @@ class SFTPClient:
         try:
             while True:
                 for entry in await self._handler.readdir(handle):
-                    if isinstance(path, str):
+                    if isinstance(path, (str, PurePath)):
                         entry.filename = self.decode(entry.filename)
                         entry.longname = self.decode(entry.longname)

@@ -3382,7 +3382,7 @@ class SFTPClient:
         if len(names) > 1:
             raise SFTPBadMessage('Too many names returned')

-        return self.decode(names[0].filename, isinstance(path, str))
+        return self.decode(names[0].filename, isinstance(path, (str, PurePath)))

     async def symlink(self, oldpath, newpath):
         """Create a remote symbolic link
```

The second change fixes a similar issue that could occur with a call to readlink().

I'm running unit tests on this now and will check the change into the "develop" branch shortly if all goes well.

**Comment by ronf** [comment]
Ok - this change is now available in the "develop" branch as commit 5cd57a1 and will be rolled into the next release. Thanks again for the report!

**Comment by Mulugruntz** [comment]
Sweet! You're very reactive, @ronf !
The tests don't check yet for when `Path` is passed though. Right?

**Comment by ronf** [comment]
From a code coverage perspective, the existing unit tests already covered passing in `str`, `bytes`, or `Path` (really any `PurePath`) values. It made sure all of these worked correctly, but that code was written to work equally well with both `bytes` and `str` values coming back in the readdir()/listdir()/scandir() calls. Code internal to AsyncSSH that used these functions was also designed to work equally well with both return types, or in some cases to convert the incoming paths to `bytes` before making these calls, since that's the more general case that can handle all types of path input.

Since the current code doesn't actually attempt to return Path() objects for the returned filenames (and I'm not sure that would make sense since these will always be simple names and not paths), I probably should update the documentation for these functions, though. Right now it says something like:

           :returns: A list of file/subdirectory names, matching the
                     type used to pass in the path```

As part of that, I could also add a unit test to explicitly check the return types for different input path types.

**Comment by ronf** [comment]
The fix for this is now available in AsyncSSH 2.6.0.
